### ISSUE TITLE
DataParallel with distributions

### ISSUE THREAD

**Issue by kamenbliznashki** [pytorch/pytorch#14329]
DataParallel throws a "zip argument must support iteration" error when the model forward returns a torch.distirbution e.g if model returns a Bernoulli and I'd like to call log_prob(inputs) to calculate my loss function. 

Is there a work around without having to drop using the distributions module?

**Comment by soumith** [comment]
DistributedDataParallel is what would work with anything non-standard neural network, such as distributions, RNNs with batch as 2nd dimension, etc.
I suggest you use that instead of DataParallel to make your life easier.

**Comment by lliimsft** [comment]
Hi,

I have a related question about Multi-GPU vs Distributed training, probably unrelated to BERT itself.

I have a 4-GPU server, and was trying to run my training in two ways:

(a) Run single-node distributed training using DistributedDataParallel with 4 processes and minibatch of N each
(b) Run Multi-GPU training with minibatch of 4N using DataParallel, and all other hyperparams keep the same

Intuitively I believe (a) and (b) should yield similar accuracy and training times. Below please find my observations:

(a) runs ~20% faster than (b).
(b) yields a better final evaluation accuracy of ~4% than (a)

The first looks like reasonable since I guess the loss.mean() is done by CPU which may be slower than using NCCL directly? However, I don't quite understand the second observation. Can you please give any hint or reference about the possible cause?

I am using ADAM optimizer FYI... 

Thanks!
